Python/Module5_OddsAndEnds/Writing_Good_Code.ipynb

```diff
@@ -0,0 +1,1028 @@
+    "- A `References` section, used to document any works that were cited in the \"Notes\" section. It is not very common to need to reference primary sources in your docstrings.\n",
+    "- An `Examples` section, which contains console-style code, similar to the code snippets that you see in PLYMI, for using your function.\n",
+    "\n",
+    "There are additional, more obscure or technical sections that you can read about in the formal specification.\n",
+    "\n",
+    "The following function has a docstring that exemplifies all of these sections.\n",
+    "\n",
+    "```python\n",
+    "def pairwise_dists(x: np.ndarray, y: np.ndarray) -> np.ndarray:\n",
+    "    \"\"\" Computes pairwise distances between the rows of ``x`` and ``y``.\n",
+    "    \n",
+    "    Returns the shape-(M, N) array of Euclidean distances between \n",
+    "    the M rows of ``x`` and the N rows of ``y``.\n",
+    "\n",
+    "    Parameters\n",
+    "    ----------\n",
+    "    x : numpy.ndarray, shape=(M, D)\n",
+    "        An optional description of ``x``\n",
+    "    y : numpy.ndarray, shape=(N, D)\n",
+    "        An optional description of ``y``\n",
+    "\n",
+    "    Returns\n",
+    "    -------\n",
+    "    numpy.ndarray, shape=(M, N)\n",
+    "        The pairwise distances\n",
+    "    \n",
+    "    Notes\n",
+    "    -----\n",
+    "    This function uses the memory-efficient vectorized implementation that\n",
+    "    is detailed in [1]_.\n",
+    "    \n",
+    "    References\n",
+    "    ----------\n",
+    "    .. [1] Soklaski, R. (2019, Jan 21). Array Broadcasting. \n",
+    "           Retrieved from https://http://www.pythonlikeyoumeanit.com\n",
+    "    \n",
+    "    Examples\n",
+    "    --------\n",
+    "    Compute the pairwise distances between the rows of a shape-(3, 3) array \n",
+    "    with the rows of a shape-(2, 3) array.\n",
+    "    \n",
+    "    >>> import numpy as np\n",
+    "    >>> x = np.array([[1., 2., 3.],\n",
+    "    ...               [4., 5., 6.],\n",
+    "    ...               [7., 8., 9.]])\n",
+    "    >>> y = np.array([[1., 2., 3.],\n",
+    "    ...               [4., 5., 6.]])\n",
+    "    >>> pairwise_dists(x, y)\n",
+    "    array([[ 0.        ,  5.19615242],\n",
+    "           [ 5.19615242,  0.        ],\n",
+    "           [10.39230485,  5.19615242]])\n",
+    "    \"\"\"\n",
+    "    ...\n",
+    "```\n",
+    "\n",
+    "See that we specify **type information** in the `Parameters` and the `Returns` section of the docstring, even though it is redundant with the function signatures' type-hints - it is useful to have this information in the docstring too. This is a good place to embellish the type information with pertinent information that is not conveyed by the formal type-hints. For example, we added type-information alongside the types in these sections. \n",
```

Comment:
signature's <-- move the apostrophe

For example, we added **shape** information alongside the types <-- we're not adding type information to the types :grimacing: